doc/design/blob-uuid-1.md

```diff
@@ -0,0 +1,45 @@
+A blob is referenced by a key generation function. the value of this key is used to index the blob in the database table of identities and on the s3 storage. `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<restofkey>`
+
+** database table of identities **: indexed by the key and contains columns for as many types of checksums as we want. these checksums constitute identities of the blob and should use functions that will always return the same value.
+
+**key generation function**: the current suggestion (from satra) for this function is UUID4.
+
+**DANDI ETag generation function**: this is an agreement between API server and DANDI CLI for now to ensure there is an algorithm that consistently returns the correct number of parts given a `contentSize` in bytes. this should in theory also be the AWS ETag generation function.
+
+## Upload (from a CLI)
+
+### Check process (for de-duplication):
+1. CLI sends expected digest with digest type (e.g., ETag)
+2. API checks for existence of ETag in identity table.
+3. if exists, returns key, if not returns an error.
+
+### Blob upload process:
+1. CLI initiates upload sending contentSize + ETag
+2. API generates key, determines upload location `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<key>`
+4. API saves initial data (size, ETAG, key) to Upload table
+3. API returns presigned URLs
+4. CLI verifies that the number+size of parts match the calculated number of parts (i.e. server and CLI are using the same ETag generation function)
+5. CLI uploads to presigned URL and checks part ETags on return. Any failures can be retried without involving API.
+6. CLI sends part info to API, API responds with presigned completion URL
+7. CLI completes upload and checks final ETag. Mismatch aborts the upload.
+8. API validates the size and that the initially reported ETag matches the actual ETag
+9. Check for collision, since some other task could have finished by this time with the same object
+10. API adds key + ETag to Asset table, delete from Upload table
+11. API kicks off background process to calculate checksums
```

Comment:
a side-note that if we stay true to current design -- addition of any new checksum should change asset UUID unless we keep those outside of the metadata records and treat them as some "dynamic" overlay ...

Reply:
I think dynamically injecting the checksums like we already do with asset path info would be better. I can imagine someone uploading some new data to a dandiset, checking that it's there in the UI, then publishing it immediately before the background checksum calculations complete.

Reply:
any new checksum is an overlay on the uuid - should not change uuid of a blob.